### ISSUE TITLE
Resume issue

### ISSUE THREAD

**Issue by Burhan-Q** [ultralytics/ultralytics#8561]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

Comparing plots (see MRE section) of run result from MRE below shows spike in loss and dip in mAP when training is resumed. 

### Environment

Reproduced results with environment:
```
Ultralytics YOLOv8.1.19 🚀 Python-3.10.12 torch-2.2.0+cu121 CUDA:0 (NVIDIA GeForce RTX 2060, 5924MiB)
Setup complete ✅ (12 CPUs, 15.6 GB RAM, 83.8/101.0 GB disk)

OS                  Linux-6.6.10-76060610-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.12
Install             pip
RAM                 15.56 GB
CPU                 AMD Ryzen 5 1600 Six-Core Processor
CUDA                12.1

matplotlib          ✅ 3.8.1>=3.3.0
opencv-python       ✅ 4.8.1.78>=4.6.0
pillow              ✅ 10.1.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.11.3>=1.4.1
torch               ✅ 2.2.0>=1.8.0
torchvision         ✅ 0.17.0>=0.9.0
tqdm                ✅ 4.66.1>=4.64.0
psutil              ✅ 5.9.6
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.1.3>=1.1.4
seaborn             ✅ 0.13.0>=0.11.0
```

### Minimal Reproducible Example

```
yolo train data=coco128.yaml model=yolov8n.pt epochs=10 exist_ok close_mosaic=5
```

Interrupt training after epoch 5 completes

```
yolo train model=runs/detect/train/weights/last.pt resume
```

### Additional

<details><summary>Testing by @Laughing-q </summary>
<p>

![image](https://github.com/ultralytics/ultralytics/assets/62214284/0d8bc50b-c44a-4d52-82f9-bf9fee3d8be8)

- blue line is training without interruption
- green line is training with `resume` before/at mosaic closing

</p>
</details> 

<details><summary>Reproduced by @Burhan-Q </summary>
<p>

![image](https://github.com/ultralytics/ultralytics/assets/62214284/a8f9268e-0ab5-4058-a559-40b157148739)

- green line is training without interruption
- blue line is training with `resume` before/at mosaic closing

</p>
</details> 

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by Burhan-Q** [comment]
Tried another round and interrupted at epoch 3. 
- blue line :: stopped at epoch 3
- green line :: stopped after epoch 5
- purple line :: no stopping

![image](https://github.com/ultralytics/ultralytics/assets/62214284/69a9cd54-5f86-470e-b992-741a0ea444a2)

**Comment by Laughing-q** [comment]
@Burhan-Q updated the name of issue since it's not related to close_mosaic process

**Comment by Laughing-q** [comment]
@Burhan-Q @glenn-jocher hey guys, after some tests I figured this might not be able to fixed...or should not be called an issue..
The reason why these two curves are different is that the batched images are different from the dataloader, when we resuming a training from a interrupted epoch, the dataloader would treat it as a new training and produce batched images just like the first epoch.
I launched a training and interrupted it manually at the beginning of the epoch 2. We have 8 iterations when training our `coco128.yaml` with default `batch=16`, so `train_batch0.jpg` is the first batched image at `epoch 1` and `train_batch8.jpg` is the first batched image at `epoch 2`. Here's training examples I got:
![image](https://github.com/ultralytics/ultralytics/assets/61612323/e6fc7cc8-abf6-46f2-881f-66b5622b4925)

We are expecting that `resumed_train_batch8.jpg` be the same as `train_batch8.jpg` but in fact it's same as `train_batch0.jpg`, because dataloader treats resuming training as a new training and it generates images from scratch, it's something happens inside the pytorch dataloader which I can not think of a way to fix it for now.

This might only affect small datasets like `coco128.yaml` as other settings(training setting and model setting) are correctly resumed and all the images would still be used to update model, just in different orders. I think training models with larger dataset could get over it.

**Comment by Laughing-q** [comment]
ok I tested a training with coco val2017, which contains 5000 images and the curves looks much closer:
![results-mosaic](https://github.com/ultralytics/ultralytics/assets/61612323/cb51a274-6c47-4190-9499-771464492c23)
(the blue one is the resuming one)

**Comment by Laughing-q** [comment]
I tried to rollback package to a very old verion like 8.0.100 and test on coco128.yaml, the two curves still went to different ways, so this is not introduced by recent updates and I'm sure it's related to the difference of batched images from dataloader. And based on the curves are getting much closer when scale up dataset to 5000 images, It's not an issue then, training model with larger dataset could get over it.

**Comment by Burhan-Q** [comment]
@Laughing-q I'm in agreement, seems like this is closable. I'll wait a few hours and if nothing else is raised about this phenomenon, then I'll close it 🚀

**Comment by Laughing-q** [comment]
@Burhan-Q @glenn-jocher closing this issue. :)
